lisa/node.py

```diff
@@ -64,59 +61,28 @@ def __init__(
         self._support_sudo: Optional[bool] = None
         self._connection_info: Optional[ConnectionInfo] = None
 
-    @staticmethod
+    @classmethod
     def create(
+        cls,
         index: int,
-        capability: schema.NodeSpace,
-        node_type: str = constants.ENVIRONMENTS_NODES_REMOTE,
-        is_default: bool = False,
+        runbook: schema.Node,
         logger_name: str = "node",
         base_log_path: Optional[Path] = None,
     ) -> Node:
-        if node_type == constants.ENVIRONMENTS_NODES_REMOTE:
-            is_remote = True
-        elif node_type == constants.ENVIRONMENTS_NODES_LOCAL:
-            is_remote = False
-        else:
-            raise LisaException(f"unsupported node_type '{node_type}'")
-        node = Node(
-            index,
-            capability=capability,
-            is_remote=is_remote,
-            is_default=is_default,
+        if not cls._factory:
+            cls._factory = subclasses.Factory[Node](Node)
+
+        node = cls._factory.create_by_runbook(
```

Comment:
```TypeError: __init__() got multiple values for argument 'index'
2021-04-23 17:39:27.117 INFO LISA.lisa 'generated_0' attached to test case 'ATest.a_test': __init__() got multiple values for argument 'index'```

so you don't even have UT for this anymore. Why try to rewrite it? I can't even test, given this bug (there might be others). We're starting to dive into a very silly/stubborn situation here. Let's use the original PR to move forward, not a new one, please :/

Reply:
The error should be caused by positioned arguments. The runbook should be first argument. I pushed a change, that disallows positioned arguments for the two methods from subclasses.

I hope to unblock your work, and keep consistent styles in the code. Let's discuss about it offline.

Reply:
We have problems here, still:

create environment.py 184:
from_runbook environment.py 305:1
load_environments environment.py 328:

With the following stack trace, in your code, what is there as node_runbook is considered a Node one, and what ends up being passed to the node's init() is something with their personal things in "delay_parsed". Super annoying, no? And it's even inconsistent, apparently: That happens from env, but not from platform. My code is consistent on all paths.

Again, why try to rewrite that here? Let's please move forward with that PR instead.

Reply:
The existing nodes doesn't need to be dealt by platform. It can be accessed immediately. The platform doesn't touch existing local or remote nodes.

Reply:
I mean there is the bug in environment: nodes:, but not in platform, if it's part of the schema of the platform. it should just work anywere

Reply:
if a node is used in platform, it should call node.create to pass in the node runbook, and get the node instance. It provides a config based node. Either the schema is in platform or environment, they should be in same schema. Let me know, if I understand it incorrectly.

Reply:
And I do that (from_existing), and I said it works. What does not work is from runbook directly to environment.

Reply:
The `Factory` discover subclasses of node, and cache it in itself (`UserDict`). When a runbook with type is passed in, it will query the cache, and find the corresponding type. Then use this type to create an new instance, and reload the runbook to the schema instance.
I tested several scenarios, runbook of local, runbook of remote, runbook of requirement, and purely generated env. It works as expected. What's the observation and how to repro? I will have a look on it.

Reply:
The original PR handles different node schema classes in schema.py, and use the `factory.create_by_type_name`. 
1. It's duplicates code like `factory.create_runbook` did. It divides the logic of loading node to two place, schema.py and environment.py.
2. It couples environment and node tightly. If a platform (like hyper-v) needs to use node without the environment. How current code helps to load node schema?

Reply:
1) I know it does. That's because that factory is meant for ```T_BASECLASS = TypeVar("T_BASECLASS", bound=BaseClassMixin)```, not the case in the context I needed. And it's very tiny method--3 lines. Not any monstrosity.
2) Can you prove that? I fail to see anything like that. Which line/file?

Finally, just plug in my changes at lisa/tests/test_environment.py and see your code fail. That needs to pass.

Reply:
1. The factory is to create an functionable instance, not for a schema class. The schema class will be used in some functional instance finally. I don't see the reason to use factory or subclasses to handle schema classes. And there is a way without copy/paste any line of code. Why needs to insistent on a copy/paste way?
2. We don't have Hyper-V code here, it's in plan. It's like how node is used in `legacy_runner.py`. The `legacy_runner` can run on local node only, so its information is hardcoded and no runbook needed. For hyper-v, it can run on current hyper-v host (in this case, no extra LISA runner machine), or operating a couple of hosts.
3. I integrated the UT. I guess the failed reason is the `runbook` argument is used, but it's original status yet. It should use `self._runbook`. I created a new class method to initialize the class, so the `runbook` can works now. I will fix the black errors continually, so the typing experience can be better.

Reply:
1. The schema is always extendable, like in notifier, platform, test cases... My point is, it doesn't need to load the original type in schema.py, it can be done when an instance, which uses this schema, is needed.
2. This point is from the ask "Can you prove that? I fail to see anything like that. Which line/file?". I thought it's asking for information about where is the code of Hyper-V.
3. Cherry-picked.